### ISSUE TITLE
Coupon codes are not working

### ISSUE THREAD

**Issue by d-den** [woothemes/woocommerce#41665]
### Prerequisites

- [ ] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When adding a (valid) coupon code based on product or product category the code is not seen as valid.

### Expected behavior

The coupon code should be accepted and the discount should be calculated.

### Actual behavior

It throws the 'error' that the coupon code is not valid for the products.

### Steps to reproduce

Create a coupon code based on a product or product category and apply the code to the shopping cart during the order process.

### WordPress Environment

The issue happens on 8.3.1. I reverted to 8.3.0 and the bug was gone, the coupon code worked as expected again.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by barryhughes** [comment]
Sorry to hear that, @d-den ... however I can't seem to replicate.

Using WooCommerce 8.3.1 I created a product within my store's "Computer" category. I then created a coupon targeting this category:

<div align="center"><img width="700" alt="coupon-code-setup" src="https://github.com/woocommerce/woocommerce/assets/3594411/27be274f-e0af-4642-ac53-43558aa21f27"></div>

I was successfully able to apply this new coupon and checkout, without experiencing any errors:

<div align="center"><img width="450" alt="order-checkout" src="https://github.com/woocommerce/woocommerce/assets/3594411/1f58ce48-e828-4077-af96-4bbc5d7799da"></div>

Can you clarify if I am missing a step, or if there are other more specific criteria needed to see the bug? Thanks!

**Comment by rodelgc** [comment]
I also couldn't replicate the issue with WooCommerce 8.3.1 @d-den. While @barryhughes tested the category-linked coupon scenario, I tested the product-linked one and both product and category-linked scenarios. I was able to use and reuse the coupon code successfully and my test orders went through with the coupon discount applied.

Attaching the [SSR](https://woo.com/document/understanding-the-woocommerce-system-status-report/) would be helpful, @d-den. It'll also be good if you'd perform a plugin and theme conflict test as mentioned [here](https://woo.com/document/how-to-test-for-conflicts).
